### DISCUSSION TITLE
Can immediately release memory space when use `Task`?

### DISCUSSION THREAD

**Discussion by Coloryr** [AvaloniaUI/Avalonia#12484] [Q&A]
When open a window and close it, the memory can't release.

https://github.com/AvaloniaUI/Avalonia/assets/26276037/ca0fc145-1476-464e-9ffd-942c29d396f4

![image](https://github.com/AvaloniaUI/Avalonia/assets/26276037/7ecef5b6-05a0-4fc7-8fdd-d089ad9c8cb7)

**Comment by Coloryr** [comment]
It look like `SixLabors.ImageSharp.Image` and `Avalonia.Media.Imaging.Bitmap` took a lot memory

**Comment by Coloryr** [comment]
When use Task, it will use more memory
```
public Task<Bitmap> Image => GetImage();

private Task<Bitmap> GetImage()
{
    return Task.Run(() =>
    {
        using var image = SixLabors.ImageSharp.Image.Load(Screenshot.Local);
        image.Mutate(p =>
        {
            p.Resize(200, 120);
        });
        using var stream = new MemoryStream();
        image.SaveAsBmp(stream);
        stream.Seek(0, SeekOrigin.Begin);
        return new Bitmap(stream);
    });
}
```

```
<Image
    Width="200"
    Height="120"
    Source="{Binding Image^, FallbackValue={x:Static app:App.LoadIcon}}"
    Stretch="UniformToFill" />
<Label HorizontalAlignment="Center" Content="{Binding Name}" />
```
![image](https://github.com/AvaloniaUI/Avalonia/assets/26276037/987b2cc2-9509-4394-9222-1158f4edb04f)

but no use Task
```
public Bitmap Image { get; init; }

using var image = SixLabors.ImageSharp.Image.Load(Screenshot.Local);
image.Mutate(p =>
{
    p.Resize(200, 120);
});
using var stream = new MemoryStream();
image.SaveAsBmp(stream);
stream.Seek(0, SeekOrigin.Begin);
Image = new Bitmap(stream);
```

```
<Image
    Width="200"
    Height="120"
    Source="{Binding Image}"
    Stretch="UniformToFill" />
```
![image](https://github.com/AvaloniaUI/Avalonia/assets/26276037/da93aa5a-fe8a-4957-9cda-1644ffccaab9)
 
It's dotnet bug?

**Reply by thevortexcloud** [reply]
I have noticed this before too. I also noticed running code in parallel also will allocate even more memory that is seemingly never released. I found some old stackoverflow posts awhile back that seemed to indicate it's intentional behaviour in the runtime as a way to reserve memory for future use. But it may have been a problem in Avalonia, I was never able to work out why it happened conclusively. It's a fairly annoying problem though. In my case I was trying to load >1000 images and it made the app use several gigabytes of RAM that I never got back even after disposing everything. But the fact running async is using more RAM is not unexpected since you are doing more things at once.

**Reply by Coloryr** [reply]
Yes, but even call `Dispose` in Task or MainThread, it also do no release memory.

**Reply by Coloryr** [reply]
`SixLabors.ImageSharp.Image` and `Avalonia.Media.Imaging.Bitmap` is same

**Reply by Coloryr** [reply]
```
private Task<Bitmap> GetImage()
{
    return Task.Run(() =>
    {
        var image = SixLabors.ImageSharp.Image.Load(Screenshot.Local);
        image.Mutate(p =>
        {
            p.Resize(200, 120);
        });
        var stream = new MemoryStream();
        image.SaveAsBmp(stream);
        
        stream.Seek(0, SeekOrigin.Begin);
        var bitmap = new Bitmap(stream);
        
        Dispatcher.UIThread.Post(() =>
        {
            image.Dispose();
            stream.Dispose();
        });
        return bitmap;
    });
}
```

**Reply by timunie** [reply]
> I have noticed this before too. I also noticed running code in parallel also will allocate even more memory that is seemingly never released. I found some old stackoverflow posts awhile back that seemed to indicate it's intentional behaviour in the runtime as a way to reserve memory for future use. But it may have been a problem in Avalonia, I was never able to work out why it happened conclusively. It's a fairly annoying problem though. In my case I was trying to load >1000 images and it made the app use several gigabytes of RAM that I never got back even after disposing everything. But the fact running async is using more RAM is not unexpected since you are doing more things at once.

most likely you don't need to display all images at once. So they can and should only be loaded when needed. Also the resolution may only be what it needs to be and not full size. Last but not least one can use some kind of Caching mechanism which holds a limited number of images and disposes any outdated data. E.g: https://learn.microsoft.com/en-us/dotnet/core/extensions/caching

**Reply by thevortexcloud** [reply]
I was paginating the images and resizing them using the decode thing specifically to avoid memory problems. It did not make much of a difference in terms of releasing allocated memory (it did however change how much memory per page was used), even with aggressively disposing and explicit garbage collecting on page changes, memory usage kept going up on page changes. But for all I know I was doing something wrong. It was my first attempt at using Avalonia. I did not try that caching thing though. 

I have however (at least partially) replicated the problem. My best guess about what is happening though is that the view somehow holds onto all the image even after they are disposed. As the images are still visible in the UI. You can fairly easily see that by doing this after all the images have loaded into the view.

```c#
 Task.Run(async () => {
            await Task.Delay(3000);
            foreach (var image in bitmaps) {
                image.Dispose();
            }
        });

```
Which makes me think the Image control itself needs to be disposed (if it even can), but I have not got that far into testing it.

**Comment by Coloryr** [comment]
```
 private async Task<Bitmap> GetImage()
 {
     using var image = SixLabors.ImageSharp.Image.Load(Screenshot.Local);
     await Task.Run(() =>
     {
         image.Mutate(p =>
         {
             p.Resize(200, 120);
         });
     });
     using var stream = new MemoryStream();
     await image.SaveAsBmpAsync(stream);

     stream.Seek(0, SeekOrigin.Begin);
     var bitmap = new Bitmap(stream);

     return bitmap;
 }
```
![K8_SC0`T(XW8C1FU2XDKD9](https://github.com/AvaloniaUI/Avalonia/assets/26276037/7a41471e-888d-43a3-8ad2-e9e391d16adc)

**Reply by timunie** [reply]
you can use a using-block which will dispose the stream earlier

```cs
using (var image = ...)
{
   ... do stuff
}
```

https://learn.microsoft.com/en-us/answers/questions/836600/using-block-vs-c-8-0-using-statement
https://www.damirscorner.com/blog/posts/20200103-BeCarefulWithUsingDeclarationsInCSharp8.html

**Reply by Coloryr** [reply]
```
private Task<Bitmap> GetImage()
{

    return Task.Run(() =>
    {
        Bitmap bitmap;
        using (var image = SixLabors.ImageSharp.Image.Load(Screenshot.Local))
        {
            image.Mutate(p =>
            {
                p.Resize(200, 120);
            });
            using (var stream = new MemoryStream())
            {
                image.SaveAsBmp(stream);

                stream.Seek(0, SeekOrigin.Begin);
                bitmap = new Bitmap(stream);
            }
            return bitmap;
        }
    });
}
```
![T4}%)VC8D$O_~65V9G S`PY](https://github.com/AvaloniaUI/Avalonia/assets/26276037/fa83c7a0-6947-4ff7-b841-39f705933bc1)

**Reply by thevortexcloud** [reply]
FYI the bitmap itself is disposable too and should probably be disposed if you can when the window closes.

**Reply by Coloryr** [reply]
It not auto dispose?

**Reply by workgroupengineering** [reply]
Try using [RecyclableMemoryStream](https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream).

**Reply by thevortexcloud** [reply]
> It not auto dispose?

In theory it will be garbage collected when there are no references left to it, which may take awhile/never hapen. If you explicitly dispose it, it may happen a lot sooner. You can register it to be auto disposed though when the viewmodel is deactivated if you are using ReactiveUI and implement `IActivatableViewModel`.

The general rule for .NET is if you can dispose something/put it in a using statement, you should. There are situations where you should not do that even if you can (EG `HttpClient`) but those are fairly rare.

**Reply by Coloryr** [reply]
> Try using [RecyclableMemoryStream](https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream).

Thanks, try this.

**Reply by Coloryr** [reply]
Anyway, if the image is do `Load` in the `Task`, It use more memory, it's not dispose problem.

**Reply by Coloryr** [reply]
![image](https://github.com/AvaloniaUI/Avalonia/assets/26276037/e519df1d-eefc-464c-8010-5d9c44d31c0a)
![image](https://github.com/AvaloniaUI/Avalonia/assets/26276037/4eca04d5-929b-40a8-92fa-f9e3682cef10)
